src/test/java/com/marklogic/spark/writer/document/WriteDocumentRowsToMarkLogicTest.java

```diff
@@ -0,0 +1,119 @@
+import com.marklogic.junit5.PermissionsTester;
+import com.marklogic.junit5.XmlNode;
+import com.marklogic.spark.AbstractIntegrationTest;
+import com.marklogic.spark.Options;
+import com.marklogic.spark.TestUtil;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Row;
+import org.apache.spark.sql.SaveMode;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import javax.xml.namespace.QName;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class WriteDocumentRowsToMarkLogicTest extends AbstractIntegrationTest {
+
+    @BeforeEach
+    void setup() {
+        TestUtil.insertTwoDocumentsWithAllMetadata(getDatabaseClient());
+    }
+
+    @Test
+    void retainMetadataFromCopiedDocuments() {
+        readTheTwoTestDocuments()
+            .write().format(CONNECTOR_IDENTIFIER)
+            .option(Options.CLIENT_URI, makeClientUri())
+            .option(Options.WRITE_URI_TEMPLATE, "/new{URI}")
+            .mode(SaveMode.Append)
+            .save();
+
+        assertCollectionSize("collection1", 4);
+        assertCollectionSize("collection2", 4);
+
+        Stream.of("/new/test/1.xml", "/new/test/2.xml").forEach(uri -> {
+            XmlNode doc = readXmlDocument(uri);
+            doc.assertElementValue("/hello", "world");
+
+            DocumentMetadataHandle metadata = readMetadata(uri);
+            verifyCollections(metadata, "collection1", "collection2");
+            verifyPropertiesAndMetadataValuesWereCopied(metadata);
+
+            PermissionsTester tester = new PermissionsTester(metadata.getPermissions());
+            assertEquals(2, tester.getDocumentPermissions().size());
+            tester.assertReadPermissionExists("spark-user-role");
+            tester.assertUpdatePermissionExists("spark-user-role");
+            tester.assertReadPermissionExists("qconsole-user");
+            assertEquals(2, tester.getDocumentPermissions().get("spark-user-role").size());
+            assertEquals(1, tester.getDocumentPermissions().get("qconsole-user").size());
+        });
+    }
+
+    @Test
+    void overrideMetadataFromCopiedDocuments() {
+        readTheTwoTestDocuments()
+            .write().format(CONNECTOR_IDENTIFIER)
+            .option(Options.CLIENT_URI, makeClientUri())
+            .option(Options.WRITE_URI_TEMPLATE, "/override{URI}")
+            // Hmm... could we do metadata values like this:
+            // .option(Options.WRITE_METADATA_VALUES, "meta1,value1,meta2,value2")
+            // .option(Options.WRITE_DOCUMENT_PROPERTIES, "{org:example}key1,value1")
+            .option(Options.WRITE_COLLECTIONS, "new1,new2")
+            .option(Options.WRITE_PERMISSIONS, "spark-user-role,read,qconsole-user,update")
+            .mode(SaveMode.Append)
+            .save();
+
+        // Make sure new docs weren't added to the test collections.
```

Comment:
Is there anything that tests the ArbitraryRowFunction?

Reply:
That's the code that has been there since the 2.0 release for writing arbitrary rows, I just pulled it out of `WriteBatcherDataWriter` since that class was getting too complicated. All of the existing 2.0 write tests exercise it.

You can also see coverage data in the Sonar report at https://sonar.progress.com/component_measures?id=marklogic_marklogic-spark-connector_AYxF-ciuhXuvzhhRmJ3v&pullRequest=142&metric=new_coverage&view=list .